blockchain/message_pool/src/msgpool/selection.rs

```diff
@@ -54,26 +72,374 @@ where
             return Ok(Vec::new());
         }
         // 0b. Select all priority messages that fit in the block
-        // TODO: Implement guess gas
-        let min_gas = 1298450;
         let (result, gas_limit) = self
             .select_priority_messages(&mut pending, &base_fee, &ts)
             .await?;
 
         // check if block has been filled
-        if gas_limit < min_gas {
+        if gas_limit < MIN_GAS {
             return Ok(result);
         }
+
         // 1. Create a list of dependent message chains with maximal gas reward per limit consumed
-        let mut chains = Vec::new();
+        let mut chains = Chains::new();
         for (actor, mset) in pending.into_iter() {
-            chains.extend(create_message_chains(&self.api, &actor, &mset, &base_fee, &ts).await?);
+            create_message_chains(&self.api, &actor, &mset, &base_fee, &ts, &mut chains).await?;
         }
 
-        let (msgs, _) = merge_and_trim(chains, result, &base_fee, gas_limit, min_gas);
+        let (msgs, _) = merge_and_trim(&mut chains, result, &base_fee, gas_limit, MIN_GAS);
         Ok(msgs)
     }
 
+    async fn select_messages_optimal(
+        &self,
+        cur_ts: &Tipset,
+        target_tipset: &Tipset,
+        ticket_quality: f64,
+    ) -> Result<Vec<SignedMessage>, Error> {
+        let base_fee = self
+            .api
+            .read()
+            .await
+            .chain_compute_base_fee(&target_tipset)?;
+
+        // 0. Load messages from the target tipset; if it is the same as the current tipset in
+        //    the mpool, then this is just the pending messages
+        let mut pending = self.get_pending_messages(&cur_ts, &target_tipset).await?;
+
+        if pending.is_empty() {
+            // FIXME: lotus returns a nil, shouldn't we return an Err?
```

Comment:
@ec2 should we?

Reply:
No. If theres no pending messages, then we select no messages. Makes sense to return empty vec.

Reply:
Alright!